keps/0008-node-heartbeat.md

```diff
@@ -0,0 +1,374 @@
+     work in a similar mode (check periodically if condition changes, but report only
+     when something changed or no status was reported for a given time) and decrease
+     its reporting frequency too.
+   - In general, we recommend to keep frequencies of NodeStatus reporting in both
+     Kubelet and NodeProblemDetector in sync (once all changes will be done) and
+     that should be reflected in [NPD documentation][].
+   - Note that reducing frequency to 1 minute already gives us almost 6x improvment.
+     It seems more than enough for any foreseeable future assuming we won’t
+     significantly increase the size of object Node.
+     Note that if we keep adding node conditions owned by other components, the
+     number of writes of Node object will go up. But that issue is separate from
+     that proposal.
+1. Increasing default frequency of NodeStatus updates may potentially break customers
+   relying on frequent Node object updates. However, in non-managed solutions,
+   customers will still be able to restore previous behavior by setting appropriate
+   flag values. Thus, changing defaults to what we recommend is the path to go with.
+
+Other notes:
+
+1. Additional advantage of using LeaderElection for that purpose would be the ability
+   to exlude it from audit profile and thus reduce the audit logs footprint.
+
+[LeaderElectionRecord]: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/leaderelection/resourcelock/interface.go#L37
+[ttl controller]: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/ttl/ttl_controller.go#L155
+[NPD documentation]: https://kubernetes.io/docs/tasks/debug-application-cluster/monitor-node-health/
+
+### Risks and Mitigations
+
+Increasing default frequency of of NodeStatus updates may potential break customers
+relying on frequent Node object updates. However, in non-managed solution, customers
+will still be able to restore previous behavior by setting appropriate flag values.
+
+## Graduation Criteria
+
+This can be immediately promoted to Beta, as the API is effectively a copy of
+already existing LeaderElectionRecord.
+
+This will be promoted to GA once it's gone a sufficient amount of time as Beta with
+no changes.
+
+## Implementation History
+
+- RRRR-MM-DD: KEP Summary, Motivation and Proposal merged
+
+## Alternatives
+
+We considered a number of alternatives, most important mentioned below.
+
+### Dedicated “heartbeat” object instead of “leader election” one
+
+Instead of introducing and using “leader election” object, we considered
+introducing a dedicated “heartbeat” object for that purpose. Apart from that,
+all the details about the solution remain pretty much the same.
+
+Pros:
+
+- Conceptually easier to understand what the object is for
+
+Cons:
+
+- Introduces a new, narrow-purpose API. Leader election is already used by other
+  components, implemented using annotations on Endpoints and ConfigMaps.
```

Comment:
I don't see this as a con. Using a leader election API object for a single actor which doesn't make use of any of the "leader" aspects of it is pretty confusing.

Reply:
I think only the name "leader election" is confusing. This is a Lease API, which is what I'd call it.

The other actor is the node controller, which takes action when the Kubelet doesn't renew its lease in time. It just doesn't acquire the lease.

Reply:
Also note that my opinion on a "node heartbeat" API hasn't changed since the original proposal (https://github.com/kubernetes/kubernetes/pull/14735) in 2015. In fact, with ~60 APIs now it's even more critical that we introduce new ones with great care.